src/persist/persist_utils.hpp

```diff
@@ -178,87 +178,113 @@ class PerThreadCircBuffer{
+            cap(_cap){
+        payloads = new padded<T>[cap];
+        pushed.ui = 0;
+        popped.ui = 0;
     }
     ~FixedCircBuffer(){
         delete(payloads);
     }
-    void push(T x){
-        int curr_tail = tail.ui.load(std::memory_order_acquire);
-        // push x into tail
-        payloads[curr_tail].ui = x;
-        // try advance tail to tail+1.
-        if (allow_wait) {
-            // busy waiting when buffer is full.
-            while(full());
-        } else {
-            if (full()){
-                errexit("FixedCircBuffer full.");
-            }
-        }
-        tail.ui.store(next(tail), std::memory_order_release);
-        size++;
-    }
     bool try_push(T x){
-        int curr_tail = tail.ui.load(std::memory_order_acquire);
-        if (full()){
+        size_t curr_pushed = pushed.ui.load(std::memory_order_acquire);
+        size_t curr_popped = popped.ui.load(std::memory_order_acquire);
+        assert(curr_pushed <= curr_popped + cap);
+        if (curr_pushed == curr_popped + cap){
             // full, return false
             return false;
         } else {
-            // push x into tail
-            payloads[curr_tail].ui = x;
-            // advance tail to tail+1.
-            tail.ui.store(next(tail), std::memory_order_release);
-            size++;
+            // push x
+            payloads[curr_pushed%cap].ui = x;
+            // advance pushed counter.
+            pushed.ui.store(curr_pushed+1, std::memory_order_release);
             return true;
         }
     }
     bool try_pop(const std::function<void(T& x)>& func){
-        int curr_head = head.ui.load(std::memory_order_acquire);
-        if (tail.ui.load(std::memory_order_acquire) == curr_head){
-            // empty.
+        size_t curr_popped = popped.ui.load(std::memory_order_acquire);
+        size_t curr_pushed = pushed.ui.load(std::memory_order_acquire);
+        assert(curr_popped <= curr_pushed);
+        if (curr_popped == curr_pushed){
+            // empty
             return false;
         }
-        func(payloads[curr_head].ui);
-        head.ui.store(next(curr_head), std::memory_order_release);
-        size--;
+        // consume the to-be-popped entry.
+        func(payloads[curr_popped%cap].ui);
+        // try to pop the next unpopped entry
+        while(!popped.ui.compare_exchange_strong(curr_popped, curr_popped+1, std::memory_order_acq_rel)){
+            size_t curr_pushed = pushed.ui.load(std::memory_order_acquire);
+            assert(curr_popped <= curr_pushed);
+            if (curr_popped == curr_pushed){
+                // empty
+                return false;
+            }
+        }
+        // successfully popped.
         return true;
     }
     bool try_pop(T& x){
-        int curr_head = head.ui.load(std::memory_order_acquire);
-        if (tail.ui.load(std::memory_order_acquire) == curr_head){
-            // empty.
+        size_t curr_popped = popped.ui.load(std::memory_order_acquire);
```

Comment:
Do we want to have assert here, just like another `try_pop`?

Reply:
Yes.